### PR TITLE
DEX-1428 send match ws updates in batch

### PR DESCRIPTION
**Pull request by JennaWestenra**

### CODE REVIEW

**Review by alxndrmkd — APPROVED** [review]

**Review by ngadiyak — APPROVED** [review]

File: dex/src/main/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActor.scala [hunk 1]
```diff
@@ -174,19 +174,21 @@ class OrderBookActor(
     processEvents(timestamp, result.events)
   }
 
-  private def processEvents(timestamp: Long, events: IterableOnce[Event]): Unit =
-    events.iterator.foreach { event =>
+  private def processEvents(timestamp: Long, events: IterableOnce[Event]): Unit = {
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActor.scala:177]:
```
  private def processEvents(timestamp: Long, events: List[Event]): Unit = {
    val changes = events.flatMap { event =>
      logEvent(event)
      // DEX-1192 docs/places-and-cancels.md
      event match {
        case event: Events.OrderExecuted => WsOrdersUpdate.from(event, timestamp).some
        case event: Events.OrderCanceled => WsOrdersUpdate.from(event).some
        case _ => none
      }
    }
    NonEmptyList.fromList(events).foreach(eventsCoordinatorRef ! OrderEventsCoordinatorActor.Command.Process(_))

    changes.map(WsInternalBroadcastActor.Command.Collect).foreach(wsInternalHandlerDirectoryRef ! _)
  }
  ````

Comment (alexander-branevskiy) [reply dex/src/main/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActor.scala:177]: rewrite to this. and convert .toList for each invocation. should look better

File: dex/src/main/scala/com/wavesplatform/dex/actors/address/AddressActor.scala [hunk 2]
```diff
@@ -153,64 +155,69 @@ class AddressActor(
...
+                    case None => throw new IllegalStateException(s"Can't find order ${remaining.id}")
+                  }
               }
-          }
-        }
-        .filterNot(_._2 == 0) // Fee could be 0 if an order executed by a small amount
+            }
+            .filterNot(_._2 == 0) // Fee could be 0 if an order executed by a small amount
+
+          val (updated, changes: AddressBalance.Changes) =
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/actors/address/AddressActor.scala:194]: explicit type here can be omitted

File: dex/src/main/scala/com/wavesplatform/dex/actors/address/AddressActor.scala [hunk 3]
```diff
@@ -914,23 +951,30 @@ object AddressActor {
     case class ObservedTxData(orders: Seq[Order], pessimisticChanges: PositiveMap[Asset, Long])
     case class MarkTxsObserved(txsWithSpending: Map[ExchangeTransaction.Id, ObservedTxData]) extends Command
 
-    sealed trait HasOrderBookEvent {
-      def event: Events.Event
+    sealed trait HasOrderBookEvents {
+      def events: Iterable[Events.Event]
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/actors/address/AddressActor.scala:955]: why we change this?

**Review by alexander-branevskiy — CHANGES_REQUESTED** [review]

File: dex/src/main/scala/com/wavesplatform/dex/actors/events/OrderEventsCoordinatorActor.scala [hunk 4]
```diff
@@ -68,44 +69,50 @@ object OrderEventsCoordinatorActor {
...
+                  acc
 
-            case event: Events.OrderCanceled =>
-              // If we here, AddressActor is guaranteed to be created, because this happens only after Events.OrderAdded
-              addressDirectoryRef ! AddressActor.Command.ApplyOrderBookCanceled(event)
-              Behaviors.same
-          }
+              }
+            }
+          NonEmptyList.fromList(orderExecutedEvents).map(nel => AddressActor.Command.ApplyOrderBookExecuted(nel)).foreach(addressDirectoryRef ! _)
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/actors/events/OrderEventsCoordinatorActor.scala:114]: ```.map(AddressActor.Command.ApplyOrderBookExecuted(_))``` or probably even ```.map(AddressActor.Command.ApplyOrderBookExecuted)```

File: dex/src/test/scala/com/wavesplatform/dex/actors/HasOecInteraction.scala [hunk 5]
```diff
@@ -12,9 +13,19 @@ trait HasOecInteraction {
 
   implicit final class TestProbeOps(val self: TestProbe) {
 
-    def expectOecProcess[T <: Events.Event](implicit ct: ClassTag[T]): T =
-      self.expectMsgType[Process].event match {
-        case event: T => event
+    def expectOecProcess[T <: Events.Event](implicit ct: ClassTag[T]): NonEmptyList[T] =
```
Comment (alexander-branevskiy) [inline dex/src/test/scala/com/wavesplatform/dex/actors/HasOecInteraction.scala:16]:
```
    def expectOecProcess[T <: Events.Event](implicit ct: ClassTag[T]): NonEmptyList[T] =
      NonEmptyList.fromListUnsafe(self.expectMsgType[Process].events.toList).map {
        case x: T => x
        case ev => fail(s"Expected ${ct.runtimeClass.getName}, given: $ev")
      }
      ```

Comment (alexander-branevskiy) [reply dex/src/test/scala/com/wavesplatform/dex/actors/HasOecInteraction.scala:16]: btw what is oec ?

File: dex/src/test/scala/com/wavesplatform/dex/actors/HasOecInteraction.scala [hunk 6]
```diff
@@ -12,9 +13,19 @@ trait HasOecInteraction {
...
+    private def loop[T <: Events.Event](events: List[Events.Event])(implicit ct: ClassTag[T]): NonEmptyList[T] =
+      events match {
+        case (xs: T) :: Nil => NonEmptyList.one(xs)
+        case (xs: T) :: tail => xs :: loop[T](tail)
+        case event => fail(s"Expected ${ct.runtimeClass.getName}, given: $event")
+      }
+
+    def expectFirstOec[T <: Events.Event](implicit ct: ClassTag[T]): T =
+      self.expectMsgType[Process].events.toList match {
+        case (head: T) :: _ => head
```
Comment (alexander-branevskiy) [inline dex/src/test/scala/com/wavesplatform/dex/actors/HasOecInteraction.scala:28]: pattern match case Nil will fail

File: dex/src/test/scala/com/wavesplatform/dex/actors/HasOecInteraction.scala (hunk 5 above)
Comment (alexander-branevskiy) [reply dex/src/test/scala/com/wavesplatform/dex/actors/HasOecInteraction.scala:16]: your variant will fail on ```case Nil```

File: dex/src/main/scala/com/wavesplatform/dex/actors/address/AddressActor.scala [hunk 7]
```diff
@@ -153,64 +155,69 @@ class AddressActor(
...
+              AddressBalance.NotObservedTxData(ownerRemainingOrders.map(_.id), NegativeMap(cumulativeDiff))
+            )
+          balances = updated
+
+          val reservedAssets = ownerRemainingOrders.flatMap(_.requiredBalance.keys).toSet
+          val newReserved = balances.reserved.filter { case (asset, _) => reservedAssets.contains(asset) }
+          log.info(
+            s"[Balance] 1. 💵: ${format(balances.tradableBalance(cumulativeDiff.keySet).xs)}; e: ${format(cumulativeDiff)}, ov: ${format(newReserved)}"
+          )
+          val currentUpdate = WsOrderExecutedUpdates(ownerRemainingOrders.map(k => AcceptedOrderWithTx(k, txResult)), changes)
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/actors/address/AddressActor.scala:206]: AcceptedOrderWithTx(_, txResult)

File: dex/src/main/scala/com/wavesplatform/dex/actors/address/AddressBalance.scala [hunk 8]
```diff
@@ -198,7 +198,7 @@ object AddressBalance {
           x.addedNotObservedTxs ++ y.addedNotObservedTxs,
           x.removedNotObservedTxs ++ y.removedNotObservedTxs,
           x.addedNotCreatedTxs ++ y.addedNotCreatedTxs,
-          x.removedNotCreatedTxs ++ x.removedNotCreatedTxs
+          x.removedNotCreatedTxs ++ y.removedNotCreatedTxs
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/actors/address/AddressBalance.scala:201]: :+1:

File: dex/src/test/scala/com/wavesplatform/dex/actors/HasOecInteraction.scala [hunk 9]
```diff
@@ -12,9 +13,19 @@ trait HasOecInteraction {
 
   implicit final class TestProbeOps(val self: TestProbe) {
 
-    def expectOecProcess[T <: Events.Event](implicit ct: ClassTag[T]): T =
-      self.expectMsgType[Process].event match {
-        case event: T => event
+    def expectOecProcess[T <: Events.Event](implicit ct: ClassTag[T]): NonEmptyList[T] =
+      loop(self.expectMsgType[Process].events.toList)
+
+    private def loop[T <: Events.Event](events: List[Events.Event])(implicit ct: ClassTag[T]): NonEmptyList[T] =
```
Comment (alexander-branevskiy) [inline dex/src/test/scala/com/wavesplatform/dex/actors/HasOecInteraction.scala:19]: will fail on pattern match case Nil

**Review by alexander-branevskiy — APPROVED** [review]

File: dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala [hunk 10]
```diff
@@ -159,8 +162,9 @@ class OrderBookActorSpecification
       tp.expectMsgType[OrderBookSnapshotUpdateCompleted]
       orderBook ! RestartActor
 
-      tp.expectOecProcess[OrderAdded] should matchTo(OrderAdded(LimitOrder(ord1), OrderAddedReason.OrderBookRecovered, ord1.timestamp))
-      tp.expectOecProcess[OrderAdded] should matchTo(OrderAdded(LimitOrder(ord2), OrderAddedReason.OrderBookRecovered, ord2.timestamp))
+      val events = tp.expectMsgType[Process].events
```
Comment (alexander-branevskiy) [inline dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala:165]:
```
val events = tp.expectMsgType[Process].events.toList
```

and then 
```
events.size shouldBe 2
events.head
events.last
```

File: dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala [hunk 11]
```diff
@@ -169,17 +173,18 @@ class OrderBookActorSpecification
...
       actor ! wrapLimitOrder(ord1)
-      tp.expectOecProcess[OrderAdded]
+      tp.expectFirstOec[OrderAdded]
 
       actor ! wrapLimitOrder(ord2)
-      tp.expectOecProcess[OrderAdded]
-      tp.expectOecProcess[OrderExecuted]
+      val events = tp.expectMsgType[Process].events
+      events.head shouldBe a[OrderAdded]
+      events.collect { case a: OrderExecuted => a }.size should be >= 1
```
Comment (alexander-branevskiy) [inline dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala:181]: can we do an exact match?

Comment (alexander-branevskiy) [reply dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala:181]: and also check expected Process size

File: dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala [hunk 12]
```diff
@@ -169,17 +173,18 @@ class OrderBookActorSpecification
...
 
       actor ! SaveSnapshot(Long.MaxValue)
       tp.expectMsgType[OrderBookSnapshotUpdateCompleted]
 
       actor ! RestartActor
-      tp.expectOecProcess[OrderAdded] should matchTo(
+      tp.expectOecProcess[OrderAdded].head should matchTo(
```
Comment (alexander-branevskiy) [inline dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala:187]: expectFirstOec

File: dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala [hunk 13]
```diff
@@ -202,14 +207,14 @@ class OrderBookActorSpecification
...
-      tp.receiveN(5)
+      tp.receiveN(3)
 
       actor ! SaveSnapshot(Long.MaxValue)
       tp.expectMsgType[OrderBookSnapshotUpdateCompleted]
       actor ! RestartActor
 
       val restAmount = ord1.amount + ord2.amount - ord3.amount
-      tp.expectOecProcess[OrderAdded] should matchTo(
+      tp.expectOecProcess[OrderAdded].head should matchTo(
```
Comment (alexander-branevskiy) [inline dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala:217]: expectFirstOec

File: dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala [hunk 14]
```diff
@@ -234,14 +239,14 @@ class OrderBookActorSpecification
...
-      tp.receiveN(7)
+      tp.receiveN(4)
 
       actor ! SaveSnapshot(Long.MaxValue)
       tp.expectMsgType[OrderBookSnapshotUpdateCompleted]
       actor ! RestartActor
 
       val restAmount = ord1.amount + ord2.amount + ord3.amount - ord4.amount
-      tp.expectOecProcess[OrderAdded] should matchTo(
+      tp.expectOecProcess[OrderAdded].head should matchTo(
```
Comment (alexander-branevskiy) [inline dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala:249]: expectFirstOec

File: dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala [hunk 15]
```diff
@@ -460,9 +465,9 @@ class OrderBookActorSpecification
         // place when order fee settings is DynamicSettings(0.001.waves, 0.005.waves)
         val taker1 = buy(wavesUsdPair, 10.waves, 3.00, matcherFee = 0.005.waves.some, ts = now.some)
         orderBook ! wrapLimitOrder(1, taker1)
-        tp.expectOecProcess[OrderAdded]
-
-        val oe = tp.expectOecProcess[OrderExecuted]
+        val events = tp.expectMsgType[Process].events
```
Comment (alexander-branevskiy) [inline dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala:468]: pls check Process expected size

File: dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala [hunk 16]
```diff
@@ -504,19 +509,20 @@ class OrderBookActorSpecification
         }
 
         orderBook ! wrapMarketOrder(marketOrder)
-        tp.expectOecProcess[OrderAdded]
-        val oe1 = tp.expectOecProcess[OrderExecuted]
+        val events = tp.expectMsgType[Process].events
```
Comment (alexander-branevskiy) [inline dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala:512]: pls check Process expected size

File: dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala [hunk 17]
```diff
@@ -565,8 +571,9 @@ class OrderBookActorSpecification
 
         withClue("Stop condition - no counter orders:") {
           orderBook ! wrapMarketOrder(marketOrder)
-          tp.expectOecProcess[OrderAdded]
-          val oc = tp.expectOecProcess[OrderCanceled]
+          val events = tp.expectMsgType[Process].events
```
Comment (alexander-branevskiy) [inline dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala:574]: pls check Process expected size

File: dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala [hunk 18]
```diff
@@ -580,13 +587,14 @@ class OrderBookActorSpecification
           tp.expectOecProcess[OrderAdded]
 
           orderBook ! wrapMarketOrder(marketOrder)
-          tp.expectOecProcess[OrderAdded]
-          val oe = tp.expectOecProcess[OrderExecuted]
+          val events = tp.expectMsgType[Process].events
```
Comment (alexander-branevskiy) [inline dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala:590]: pls check Process expected size

File: dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala [hunk 19]
```diff
@@ -637,12 +645,15 @@ class OrderBookActorSpecification
...
 
           orderBook ! wrapLimitOrder(counterOrder)
-          tp.expectOecProcess[OrderAdded]
+          tp.expectFirstOec[OrderAdded]
 
           orderBook ! wrapMarketOrder(marketOrder)
 
-          tp.expectOecProcess[OrderAdded]
-          val oe = tp.expectOecProcess[OrderExecuted]
+          val events = tp.expectMsgType[Process].events
```
Comment (alexander-branevskiy) [inline dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala:652]: pls check Process expected size

File: dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala [hunk 20]
```diff
@@ -724,7 +740,7 @@ class OrderBookActorSpecification
       tp.expectOecProcess[OrderAdded]
 
       orderBook ! wrapCommand(ValidatedCommand.DeleteOrderBook(wctWavesPair))
-      tp.expectOecProcess[OrderCanceled].reason shouldBe Events.OrderCanceledReason.OrderBookDeleted
+      tp.expectOecProcess[OrderCanceled].head.reason shouldBe Events.OrderCanceledReason.OrderBookDeleted
```
Comment (alexander-branevskiy) [inline dex/src/test/scala/com/wavesplatform/dex/actors/orderbook/OrderBookActorSpecification.scala:743]: expectFirstOec

**Pull request merged**